src/piece_picker.cpp

```diff
@@ -3470,7 +3535,7 @@ namespace {
 			if (i->finished < blocks_in_piece(i->index))
 				return;
 
-			if (i->passed_hash_check)
+			if (i->passed_hash_check && i->hashing == 0)
```

Comment:
is it possible for `passed_hash_check` to be true while `i->hashing` > 0?
I have a feeling that maybe this should be an assert instead.

Reply:
it seems that if this can happen, there need to be a check when we decrement `hashing` as well, to maybe call `we_have()` at that point.